client/src/api/block_builder/input_selection/core/mod.rs

```diff
@@ -218,35 +216,41 @@ impl InputSelection {
         self
     }
 
-    // TODO should we somehow enforce using filter so we don't have to use can_be_unlocked_now later everywhere ?
-    /// Filters out the available inputs that
-    /// - can't be unlocked by the given addresses
-    /// - can't be unlocked now
-    pub fn filter(self, addresses: &[Address]) -> Self {
-        let _addresses = addresses
-            .iter()
-            // TODO meh
-            .copied()
-            .chain(self.available_inputs.iter().filter_map(|input| match &input.output {
-                Output::Alias(output) => Some(Address::Alias(AliasAddress::from(
-                    output.alias_id_non_null(input.output_id()),
-                ))),
-                Output::Nft(output) => Some(Address::Nft(NftAddress::from(
-                    output.nft_id_non_null(input.output_id()),
-                ))),
-                _ => None,
-            }));
-
-        // self.available_inputs.retain(|input| input.can_be_unlocked_now(input, addresses, ???, self.time))
+    fn filter_inputs(&mut self) {
+        self.available_inputs.retain(|input| {
+            if !input.output.is_basic()
+                && !input.output.is_alias()
+                && !input.output.is_foundry()
+                && !input.output.is_nft()
+            {
+                return false;
+            }
 
-        self
+            // PANIC: safe to unwrap as non basic/alias/foundry/nft outputs are already filtered out.
+            let unlock_conditions = input.output.unlock_conditions().unwrap();
+
+            if unlock_conditions.is_time_locked(self.timestamp) {
+                return false;
+            }
+
+            let required_address = input
+                .output
+                .required_and_unlocked_address(self.timestamp, input.output_id(), true)
+                // PANIC: safe to unwrap as non basic/alias/foundry/nft outputs are already filtered out.
+                .unwrap()
+                .0;
+
+            self.addresses.contains(&required_address)
```

Comment:
Does this mean that one has to provide also the alias and NFT outputs, if one wants to use outputs that are controlled by these?

Reply:
Ok, they will be added in L 161